lib/sandthorn/event_sourced_base.rb

```diff
@@ -0,0 +1,230 @@
+module Sandthorn
+  module EventSourced
+    module Base
+
+      attr_reader :id
+      attr_reader :events
+      attr_reader :current_event_version
+      attr_reader :originating_version
+      attr_reader :aggregate_stored_serialized_object #USED?
+      attr_reader :trace_information
+
+      alias :aggregate_id :id
+
+      def base_initialize
+        @current_event_version = 0
+        @originating_version = 0
+        @events = []
+      end
+
+      def save
+        events.each do |event|
+          event[:event_data] = Sandthorn.serialize event[:event_args]
+          event[:event_args] = nil #Not send extra data over the wire
+        end
+
+        unless events.empty?
+          Sandthorn.save_events(
+            events,
+            id,
+            self.class
+          )
+          @events = []
+          @originating_version = @current_event_version
+        end
+
+        self
+      end
+
+      def ==(other)
+        other.respond_to?(:id) && id == other.id
+      end
+
+      def trace args
+        @trace_information = args
+        yield self if block_given?
+        @trace_information = nil
+      end
+
+      def commit *args
+        attribute_deltas = get_delta
+
+        unless attribute_deltas.empty?
+          method_name = caller_locations(1,1)[0].label.gsub(/block ?(.*) in /, "")
+          increase_current_version!
+
+          data = {
+            method_name: method_name,
+            method_args: args,
+            attribute_deltas: attribute_deltas
+          }
+          trace_information = @trace_information
+          unless trace_information.nil? || trace_information.empty?
+            data.merge!({ trace: trace_information })
+          end
+
+          @events << ({
+            aggregate_version: @current_event_version,
+            event_name: method_name,
+            event_args: data
+          })
+        end
+
+        self
+      end
+
+      alias :record_event :commit
+
+      module ClassMethods
+
+        @@event_sourced_attributes = []
```

Comment:
Is this use of class variables intended? Subclasses share the same class variables, so you can get very interesting bugs in this way. For example: if you create a subclass and then run `event_sourced_attributes` again, you will mutate what attributes the superclass will event source.